src/transformers/models/lightglue/modeling_lightglue.py

```diff
@@ -0,0 +1,1158 @@
+        self.output = LightGlueSelfOutput(config)
+
+    def forward(
+        self,
+        hidden_states,
+        attention_mask=None,
+        sinusoidal_pos=None,
+        head_mask=None,
+        encoder_hidden_states=None,
+        encoder_attention_mask=None,
+        past_key_value=None,
+        output_hidden_states=False,
+        output_attentions=False,
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]], Optional[Tuple[torch.Tensor]]]:
+        self_outputs = self.self(
+            hidden_states,
+            attention_mask,
+            sinusoidal_pos,
+            head_mask,
+            encoder_hidden_states,
+            encoder_attention_mask,
+            past_key_value,
+            output_attentions,
+        )
+        self_output = self_outputs[0]
+        attention_probs = self_outputs[1:] if output_attentions else None
+        outputs = self.output(self_output, hidden_states, output_hidden_states=output_hidden_states)
+        output = outputs[0]
+        hidden_states = outputs[1] if output_hidden_states else None
+        outputs = (output, hidden_states, attention_probs)
+        return outputs
+
+
+class LightGlueTransformerLayer(nn.Module):
+    def __init__(self, config: LightGlueConfig):
+        super().__init__()
+        self.self_attention_block = LightGlueAttention(config)
+        self.cross_attention_block = LightGlueAttention(config)
+
+    def forward(
+        self,
+        descriptors: torch.Tensor,
+        keypoints: torch.Tensor,
+        attention_mask: torch.Tensor,
+        output_hidden_states: Optional[bool] = False,
+        output_attentions: Optional[bool] = False,
+    ) -> Tuple[torch.Tensor, Optional[Tuple[torch.Tensor]], Optional[Tuple[torch.Tensor]]]:
+        all_hidden_states = () if output_hidden_states else None
+        all_attentions = () if output_attentions else None
+
+        if output_hidden_states:
+            all_hidden_states = all_hidden_states + (descriptors,)
+
+        batch_size, num_keypoints, descriptor_dim = descriptors.shape
+        # (batch_size, num_keypoints, descriptor_dim) -> (batch_size, 1, num_keypoints, descriptor_dim)
+        keypoints = keypoints.unsqueeze(-3)
+        self_attention_output = self.self_attention_block(
+            descriptors,
+            sinusoidal_pos=keypoints,
+            attention_mask=attention_mask,
+            output_hidden_states=output_hidden_states,
+            output_attentions=output_attentions,
+        )
+
+        descriptors = self_attention_output[0]
+
+        encoder_hidden_states = (
+            descriptors.reshape(-1, 2, num_keypoints, descriptor_dim)
+            .flip(1)
+            .reshape(batch_size, num_keypoints, descriptor_dim)
+        )
+        encoder_attention_mask = (
+            attention_mask.reshape(-1, 2, 1, 1, num_keypoints).flip(1).reshape(batch_size, 1, 1, num_keypoints)
+            if attention_mask is not None
+            else None
+        )
```

Comment:
Would appretiate comments here on why we are doing flip/reshape

Reply:
Comments added !